src/main/scala/org/allenai/pipeline/Producer.scala

```diff
@@ -171,6 +190,44 @@ class PersistedProducer[T, -A <: Artifact](
   override def stepInfo = step.stepInfo.copy(outputLocation = Some(artifact.url))
 }
 
+case class Status(status: String)
+
+/** Wraps a Producer into a persistence layer, which will write the disk on first execution,
+  * then read cached data from disk on subsequent executions.
+  *
+  * We ensure writes fully complete by adding a meta file to each write, the presence of
+  * which guarantees a fully complete data write.
+  *
+  * path is treated as a directory, all meta and data files will reside within the path directory.
+  */
+// REVIEW What is the value / use-case of having Producers that are not persisted?
+// Can we merge Producer + PersistedProducer and always persist the data?
+class PersistedProducer2[T](
+    step: Producer[T],
+    path: String,
+    metaFs: FlatFileSystem,
+    partialData: PartialFileItem[T]
+) extends Producer[T] {
+  def create: T = {
+    // See https://github.com/allenai/s2-offline/blob/cf0998aa1bda08d77fede1f2eb77067e70a56c3b/pipeline/src/main/scala/org/allenai/scholar/pipeline/spark/S3SequenceFileArtifact.scala#L20.
+    implicit val metaFormat = IoHelpers.asStringSerializable(jsonFormat1(Status.apply))
```

Comment:
I don't think the logic for how to represent a successful write belongs in this class.  It should be in the `FileItem` representation.